src/mame/machine/nes_vt32_soc.cpp

```diff
@@ -0,0 +1,115 @@
+// license:BSD-3-Clause
+// copyright-holders:David Haywood
+
+// used by fcpocket, dgun2573, rminitv
+
+#include "emu.h"
+#include "nes_vt32_soc.h"
+
+// these have RGB12 output mode
+DEFINE_DEVICE_TYPE(NES_VT32_SOC,     nes_vt32_soc_device,     "nes_vt32_soc", "VT32 series System on a Chip (FP) (NTSC)")
+DEFINE_DEVICE_TYPE(NES_VT32_SOC_PAL, nes_vt32_soc_pal_device, "nes_vt32_soc_pal", "VT32 series System on a Chip (FP) (PAL)")
+
+nes_vt32_soc_device::nes_vt32_soc_device(const machine_config& mconfig, device_type type, const char* tag, device_t* owner, uint32_t clock) :
+	nes_vt09_soc_device(mconfig, type, tag, owner, clock)
+{
+}
+
+nes_vt32_soc_device::nes_vt32_soc_device(const machine_config& mconfig, const char* tag, device_t* owner, uint32_t clock) :
+	nes_vt32_soc_device(mconfig, NES_VT32_SOC, tag, owner, clock)
+{
+}
+
+nes_vt32_soc_pal_device::nes_vt32_soc_pal_device(const machine_config& mconfig, const char* tag, device_t* owner, uint32_t clock) :
+	nes_vt32_soc_device(mconfig, NES_VT32_SOC_PAL, tag, owner, clock)
+{
+}
+
+void nes_vt32_soc_device::device_add_mconfig(machine_config& config)
+{
+	nes_vt02_vt03_soc_device::device_add_mconfig(config);
+
+	M6502_VTSCR(config.replace(), m_maincpu, NTSC_APU_CLOCK); // are these later chips N2A03 core, or 6502 core derived?
+	m_maincpu->set_addrmap(AS_PROGRAM, &nes_vt32_soc_device::nes_vt_fp_map);
+}
+
+void nes_vt32_soc_pal_device::device_add_mconfig(machine_config& config)
+{
+	nes_vt32_soc_device::device_add_mconfig(config);
+	do_pal_timings_and_ppu_replacement(config);
+}
+
+
+uint8_t nes_vt32_soc_device::vtfp_4119_r()
+{
+	// would be PAL/NTSC etc. in base system, maybe different here?
+	return 0x00;
+}
+
+void nes_vt32_soc_device::vtfp_411e_w(uint8_t data)
+{
+	logerror("411e_w %02x\n", data);
+	if (data == 0x05)
+		dynamic_cast<m6502_vtscr&>(*m_maincpu).set_next_scramble(true);
+	else if (data == 0x00)
+		dynamic_cast<m6502_vtscr&>(*m_maincpu).set_next_scramble(false);
```

Comment:
These should use `downcast` if possible to avoid the overhead of a dynamic cast in release builds.

Reply:
this isn't new code, it's just moved code, it's already like this in the source right now.